src/libexpr/flake/call-flake.nix

```diff
@@ -43,7 +43,37 @@ let
                 (resolveInput lockFile.nodes.${nodeName}.inputs.${builtins.head path})
                 (builtins.tail path);
 
-          outputs = flake.outputs (inputs // { self = result; });
+          # Attributes that are added to the final representation of the flake.
+          # NB: `attrNames extraAttributes` must be lazy in `outputs` (tested). Values may depend on `outputs`.
+          extraAttributes =
+            sourceInfo
+            // {
+              # This shadows the sourceInfo.outPath
+              inherit outPath;
+
+              inherit inputs; inherit outputs; inherit sourceInfo; _type = "flake";
+            };
+
+          meta = {
+            # The source root, which may not correspond to the flake directory.
+            inherit sourceInfo;
+            # The base directory of the flake
+            inherit subdir;
+            # Extra attributes in the final representation of the flake, added to result of the output function
+            inherit extraAttributes;
+            # Extra inputs to the output function
+            inherit extraArguments;
+          };
+
+          # NB: `attrNames arguments` must be lazy in `outputs` (tested).
+          arguments = inputs // extraArguments;
+
+          extraArguments = {
+            self = result;
+            inherit meta;
```

Comment:
Could you inspect `flake.outputs` arguments here and only provide `meta` when it accepts it?

Reply:
Yes, most of the time, after adding more function reflection primops.

It will warn when there's a potential conflict. In `args@{ self }`, `args` won't have `.meta` because I don't think we should break function call semantics just for this.

The specifics of the conflict:
 - `args@` should always contain `meta`, or at least this would seem like a reasonable assumption.
 - `args@{ foo }` should constrain `args` to `{ foo }`, because there's no `...`. So no `meta`.

The latter requirement corresponds to actual language semantics, and backward compatibility, so that's what we pick, but we do warn about it because of the other idea.